docs/strings.rst

```diff
@@ -3,164 +3,68 @@
-whole chapter:
-
-* Use ``numpy.string_`` for scalar attributes
-* Use the NumPy ``S`` dtype for datasets and array attributes
-
-
-Fixed-length ASCII
-^^^^^^^^^^^^^^^^^^
-
-These are created when you use ``numpy.string_``:
-
-    >>> dset.attrs["name"] = numpy.string_("Hello")
-
-or the ``S`` dtype::
-
-    >>> dset = f.create_dataset("string_ds", (100,), dtype="S10")
-
-In the file, these map to fixed-width ASCII strings.  One byte per character
-is used.  The representation is "null-padded", which is the internal
-representation used by NumPy (and the only one which round-trips through HDF5).
-
-Technically, these strings are supposed to store `only` ASCII-encoded text,
-although in practice anything you can store in NumPy will round-trip.  But
-for compatibility with other programs using HDF5 (IDL, MATLAB, etc.), you
-should use ASCII only.
-
 .. note::
 
-    This is the most-compatible way to store a string.  Everything else
-    can read it.
-
-Variable-length ASCII
-^^^^^^^^^^^^^^^^^^^^^
-
-These are created when you assign a byte string to an attribute::
-
-    >>> dset.attrs["attr"] = b"Hello"
-
-or when you create a dataset with an explicit ascii string dtype::
-
-    >>> dt = h5py.string_dtype(encoding='ascii')
-    >>> dset = f.create_dataset("name", (100,), dtype=dt)
-
-Note that they're `not` fully identical to Python byte strings.  You can
-only store ASCII-encoded text, without NULL bytes::
-
-    >>> dset.attrs["name"] = b"Hello\x00there"
-    ValueError: VLEN strings do not support embedded NULLs
-
-In the file, these are created as variable-length strings with character set
-H5T_CSET_ASCII.
-
-
-Variable-length UTF-8
-^^^^^^^^^^^^^^^^^^^^^
+   The rules around reading & writing string data were redesigned for h5py
+   3.0. Refer to `the h5py 2.10 docs <https://docs.h5py.org/en/2.10.0/strings.html>`__
+   for how to store strings in older versions.
 
-These are created when you assign a unicode string to an attribute::
+Reading strings
+---------------
 
-    >>> dset.attrs["name"] = "Hello"
+String data in HDF5 datasets is read as bytes by default: ``bytes`` objects
+for variable-length strings, or numpy bytes arrays (``'S'`` dtypes) for
+fixed-length strings. Use :meth:`.Dataset.asstr` to retrieve ``str`` objects.
 
-or if you create a dataset with an explicit string dtype:
+Variable-length strings in attributes are read as ``str`` objects. These are
+decoded as UTF-8 with surrogate escaping for unrecognised bytes.
 
-    >>> dt = h5py.string_dtype()
-    >>> dset = f.create_dataset("name", (100,), dtype=dt)
+Storing strings
+---------------
 
-They can store any character a Python unicode string can store, with the
-exception of NULLs.  In the file these are created as variable-length strings
-with character set H5T_CSET_UTF8.
+When creating a new dataset or attribute, Python ``str`` or ``bytes`` objects
+will be treated as variable-length strings, marked as UTF-8 and ASCII respectively.
```

Comment:
I do not recall why we can not use this flag to also mark the bytes as `UTF-32`.

Reply:
HDF5 has no UTF-32 option (just ASCII and UTF-8, it's effectively an enum), or do you mean we convert from UTF-32 to UTF-8 and store that?

Reply:
Yup, it's a fixed list of 2 options, rather than an arbitrary tag - see [H5Tset_cset](https://confluence.hdfgroup.org/display/HDF5/H5T_SET_CSET).